### ISSUE TITLE
Unit deselecting bug

### ISSUE THREAD

**Issue by EoZ8** [wesnoth/wesnoth#2905]
Wesnoth 1.13.13

When right clicking when having a unit selected the game deselects the unit AND opens the right click menu, happens when clicking on any unit including the selected one, does not happen when clicking on empty hexes.

**Comment by ProditorMagnus** [comment]
While it it changed behavior and not ideal, I dont think this should be called bug.

**Comment by Vultraz** [comment]
It doesn't matter if it's changed. It matters if it's not ideal. I don't see it as not ideal.

**Comment by soliton-** [comment]
What‘s the idea behind the new behaviour? It doesn‘t seem intuitive to me.

**Comment by Vultraz** [comment]
I don't remember if it was intentional.

**Comment by soliton-** [comment]
Intentional or not to call this ideal or not-a-bug there should be some rationale behind it.

**Comment by ProditorMagnus** [comment]
This has not been resolved.

**Comment by Wedge009** [comment]
Hasn't it? Which version are you testing?

**Comment by Wedge009** [comment]
I understood this to be a duplicate of #1908, which has been reported resolved. Is this supposed to be different?

**Comment by ProditorMagnus** [comment]
#1908 sounds like it opens menu everywhere - this issue is only about menu opening on unit.

**Comment by ProditorMagnus** [comment]
1.14.0

**Comment by Wedge009** [comment]
Personally I'm not sure why you'd want to specifically cancel a move by right-clicking on the unit itself rather than an unoccupied hex. I also don't think it's 'not ideal' to treat a right-click on the unit as a command to bring up the context menu. Nonetheless, I concede it's different (but obscure) case to #1908. Re-opening.

**Comment by Wedge009** [comment]
@ProditorMagnus I just retested in 1.14.17 and 1.15.17+dev - doesn't seem like an issue to me. Can you check, please?

**Comment by ProditorMagnus** [comment]
Happens with 1.15.15.

**Comment by Wedge009** [comment]
Just so I'm clear (testing on 1.15.17+dev):

1. Left-click unit to select
2. Right-click elsewhere to cancel selection
3. Issue is context menu opens...?

Step 3 doesn't happen for me - just reconfirmed this on 1.15.17+dev. If it's still happening for you on the latest 1.16 branch, which OS are you using?

**Comment by rodolphecombe** [comment]
IMHO, the correct behavior when right-clicking on an unit is to open the context menu for this unit and to deselect the currently selected unit if it is another unit. Which afaict is more or less the current behavior.

**Comment by Wedge009** [comment]
It seems like that is becoming the consensus, especially in light of objection to #11368, so unless @EoZ8 responds we should probably close this.

**Comment by CelticMinstrel** [comment]
Maybe the correct thing to do here is to just remove the mechanic that right-clicking deselects a unit? I can't think of any good way to prevent that from conflicting with the context menu, which in theory needs to be available at any time in any context, whether or not a unit is currently selected. Still, as long as right-clicking _on the selected unit_ opens the context menu and does not deselect it, I think that should be good enough. Is that true currently? Given the age of the issue, I would suspect so, but I haven't tested.

Of course, if you remove the mechanic that right-clicking deselects a unit, then there needs to be a new way to do it, and I don't have any ideas on how that could work…

**Comment by ProditorMagnus** [comment]
Left click on empty hex could deselect.

**Comment by gfgtdf** [comment]
It's really unclear to me what "selcted" means in the cotext of this bug report. Is it the "Show reachmap and next leftclick moves the unit" mode or just the "There exists a unit that is shown on the right-side panel if the mouse is not hovering over another unit" mode?

**Comment by ProditorMagnus** [comment]
Good point, so left click is not possible since it has meaning for movement. Now it sounds like issue that may need more testing and this late into 1.19 might not be good time anymore.

**Comment by Wedge009** [comment]
> Maybe the correct thing to do here is to just remove the mechanic that right-clicking deselects a unit?

Unselecting a unit is pretty critical, unless you were just considering moving the key that does it not removing it altogether. I would very much prefer just closing this request rather than trying to change something so fundamental.

**Comment by babaissarkar** [comment]
What about clicking a selected unit again to deselect or double click or something like alt+click or ctrl+click?

Clicking again is how it's implemented in Android. One click: selected. Click again on that unit: deselect. Admitted, in Android you get context menu by long press, which is customary there but not on desktop, and this conflict never arises because deselection and context are assigned to different actions.

**Comment by CelticMinstrel** [comment]
> Unselecting a unit is pretty critical, unless you were just considering moving the key that does it not removing it altogether.

Yes, of course. We can't remove the ability to deselect a unit – it would make the game completely unplayable.

> What about clicking a selected unit again to deselect or double click or something like alt+click or ctrl+click?

This feels a bit unintuitive. It's not common (at least on desktop) for clicking the selected thing to deselect it (though it's not unheard of). But it I guess it could be a workable approach.

There's also the possibility of clicking with a modifier, for example shift-click or alt-click deselects… and/or adding a keyboard shortcut to deselect the current unit.

**Comment by Wedge009** [comment]
Also consider the history of Wesnoth, I have a recollection that we've had changes to fundamental game actions like these, and they're always met with resistance - this issue itself being an example of that.

Given the concerns raised over this, I closed my PR implementing this request.

**Comment by babaissarkar** [comment]
The request should be rephrased: "Avoid clash between two different right click actions".

Since how widely used context menu is, it should stay on Right click, which is the standard anyways for context menu triggers.
Deselect is much less used on desktop, so changing the shortcut for that makes more sense. I doubt the community would object given the present situation makes Right click deselection problematical to use anyways.

**Comment by babaissarkar** [comment]
> This feels a bit unintuitive. It's not common (at least on desktop) for clicking the selected thing to deselect it (though it's not unheard of). But it I guess it could be a workable approach.

I think clicking again is the more intuitive option compared to modifier + click. I don't know if this is because of MacOS behaviour differences, ~~but as an example, on Linux or Windows, I can click on file > selects it on file manager, I click again > deselect~~. Another point is that it doesn't require a modifier and the code already exists for Android, just needs to be rewired.

**Comment by soliton-** [comment]
> I don't know if this is because of MacOS behaviour differences, but as an example, on Linux or Windows, I can click on file > selects it on file manager, I click again > deselect.

On what file managers? In my experience clicking again on a selected file lets you rename it on Linux, Windows and MacOS.

I don't know why a new behavior is needed when we can simply go back to the previous one as suggested here. Deselect with right click without special behavior depending on what you right click on.

**Comment by babaissarkar** [comment]
> On what file managers? In my experience clicking again on a selected file lets you rename it on Linux, Windows and MacOS.

Probably I'm misremembering something, nvm. clicking empty area seems to deselect a file.

> I don't know why a new behavior is needed when we can simply go back to the previous one as suggested here. Deselect with right click without special behavior depending on what you right click on.

Which comment are you referring to? I don't understand given I'm somewhat late to the discussion.

**Comment by Wedge009** [comment]
Presumably the suggestion to go back to the 'old' behaviour - which is really old now, and there's been enough time for this 'new' behaviour to become entrenched such that people are using it in their custom campaigns - see comments in #11368.

**Comment by soliton-** [comment]
> Presumably the suggestion to go back to the 'old' behaviour - which is really old now, and there's been enough time for this 'new' behaviour to become entrenched such that people are using it in their custom campaigns - see comments in [#11368](https://github.com/wesnoth/wesnoth/pull/11368).

Would be good to get a clear case in any addon or mainline. "I think something somewhere might break" is not particularly helpful. If there is an expectation that `needs_select` is somehow related then the documentation needs an update.

**Comment by Wedge009** [comment]
It's not 'I think' - according to @CelticMinstrel it sounds to me like it _will_ break.

**Comment by soliton-** [comment]
Ok, I can only see guesses. Maybe someone can present an actual case.

**Comment by Wedge009** [comment]
Acknowledging there's a strong interest in seeing Wesnoth on touch devices, I consider the left and right mouse buttons fundamental to the core Wesnoth experience. So I had a look back as far as 1.10 (I could go further) and found that - at least when considering left- and right-clicks, they are generally the same across 1.10, 1.12, 1.14, 1.16, 1.18, except for the transition between 1.12 and 1.14 where at least two changes occurred, possibly due to the same changes:

* What this issue complains of - that a right-click on a unit while a unit is selected no longer just unselects the unit but also shows the context menu.
* Previously the context menu could be dismissed with a second right-click. With 1.14 onwards, the context menu remains until a selection is made or a left-click outside of the menu dismisses it.

Personally I don't feel affected by either this issue or the context menu dismissal. I seldom use the context menu, so I don't feel too concerned about it being bound to another key, however, in other applications it seems universal that a right-click brings up a context menu of some sort.

On the other hand, I think right-click unselecting a unit in Wesnoth is fundamental to its core experience and I definitely would be put off by having that re-assigned. I sense were are collectively at an impasse and I think the least disruptive action is to just let things remain as they are currently.

**Comment by soliton-** [comment]
> On the other hand, I think right-click unselecting a unit in Wesnoth is fundamental to its core experience and I definitely would be put off by having that re-assigned. I sense were are collectively at an impasse and I think the least disruptive action is to just let things remain as they are currently.

I'm a little confused why there is an impasse when you just cited how in previous versions it worked fine as: "Previously the context menu could be dismissed with a second right-click."

Is this about there being an unspecified breakage of some kind when going back to what worked just fine for years?

-----

I'm not particularly affected personally either and it's not that difficult to just not right click on units but it also seems trivial to fix to have consistent behavior as before.

**Comment by Wedge009** [comment]
I won't speak to that since you're asking for 'actual cases'. But I was referring to the new framing of 'two actions assigned to right-click by default causing a clash' - on the specific point of the context menu appearing over a unit, I really don't mind either way, but all I was trying to do was to close old issues by submitting #11368. Instead I seem to have sparked a big debate.

**Comment by babaissarkar** [comment]
I still don't understand what is the exact change @soliton- is proposing. Please directly write it here to have everything in one place.

I'd continue to request that right click context menu be not changed. That would be extremely counter intuitive to what Wesnoth and other apps generally do.

**Comment by soliton-** [comment]
If #11368 works it seems fine. I would have rather expected adjusting
https://github.com/wesnoth/wesnoth/blob/16b674cc525290954ce74a6641e8dd4db8ae5573/src/mouse_events.cpp#L858-L859

to not check `valid()`. But that is just a guess.

> But I was referring to the new framing of 'two actions assigned to right-click by default causing a clash'

That is the core issue though? The old behavior did one thing at a time and that seems simple and clear without exceptions based on where you click.

> Instead I seem to have sparked a big debate.

I see only speculation on how things might (not) work to be honest.

**Comment by babaissarkar** [comment]
> Previously the context menu could be dismissed with a second right-click."

I don't see anything odd with supporting this specifically, but how does it relate to right click doing two different things?

If you want right click doing one thing, sure. As long as that is "showing the context menu when on something that could show a context menu."

**Comment by babaissarkar** [comment]
This could probably work as an example:
My campaign, Journey of a Frost Mage has a equipment system. You can right click on a unit to show a specific Inventory option in the context menu which allows you to see what equipment the Unit is carrying or equipped to. Only happens if the hex isn't empty.

Several other campaigns maintained by @knyghtmare also use the same mechanic.

LotI also has such a mechanism that allows to show special unit stats IIRC.

**Comment by Wedge009** [comment]
Right-clicking on a unit in itself is still available even after #11368. It's the right-clicking on a unit _while a unit is selected_ that's the issue that was originally raised in the first post.

> If you want right click doing one thing, sure. As long as that is "showing the context menu when on something that could show a context menu."

I would object to the 'doing one thing', I would want to keep the right-click for unselecting units.

**Comment by soliton-** [comment]
Maybe I don't get what is meant with two actions. I took it as the two actions that are mentioned in the first post ("deselects the unit AND opens the right click menu"). If the idea is to have right click only open the context menu, i.e. completely changing how right click works in wesnoth since forever then I agree that's not a good idea.

**Comment by babaissarkar** [comment]
> would object to the 'doing one thing', I would want to keep the right-click for unselecting units.

Then it probably have to be deselect the unit + show context menu of it has something relevant, without causing a mass UMC rewrite.



> If the idea is to have right click only open the context menu, i.e. completely changing how right click works in wesnoth since forever then I agree that's not a good idea.

Well, I have no problems with keeping the present behaviour which is do two things. This issue should be closed as "won't fix" then.

**Comment by soliton-** [comment]
> Then it probably have to be deselect the unit + show context menu of it has something relevant, without causing a mass UMC rewrite.

What mass UMC rewrite? Are you implying there was a mass UMC rewrite after the behavior change (for the specific case of right clicking on a unit while another is selected as is the only thing this issue is about) of deselecting and opening the context menu? What would there even have been rewritten? Documentation that explains what happens on right clicking while a unit is selected?

**Comment by babaissarkar** [comment]
> the behavior change (for the specific case of right clicking on a unit while another is selected as is the only thing this issue is about) of deselecting and opening the context menu?

Let's clear this first. Perhaps there's a misunderstanding.
1. One unit is selected.
2. You right click.
3. Where? On that selected unit or on another unit?
4. What's happening now?
5. What is the suggested behavior?
The original is still too vague for me still.

**Comment by Wedge009** [comment]
To answer 3 and 4: I described the differences in https://github.com/wesnoth/wesnoth/pull/11368#issuecomment-4976954369

Current 1.18 behaviour (and 1.16 + 1.14):
* No unit selected + right-click anywhere -> context menu
* Unit selected + right-click anywhere except a hex with a unit -> unselect the unit
* Unit selected + right-click on a unit -> unselect unit and **show context menu** (what this issue is complaining about)

After #11368 (and behaviour for 1.12 and 1.10 and likely earlier):
* No unit selected + right-click anywhere -> context menu
* Unit selected + right-click anywhere except a hex with a unit -> unselect the unit
* Unit selected + right-click on a unit -> unselect unit (ie **No more context menu**)

**Comment by babaissarkar** [comment]
> To answer 3 and 4: I described the differences in [#11368 (comment)](https://github.com/wesnoth/wesnoth/pull/11368#issuecomment-4976954369)
> 
> Current 1.18 behaviour (and 1.16 + 1.14):
> 
>     * No unit selected + right-click anywhere -> context menu
> 
>     * Unit selected + right-click anywhere except a hex with a unit -> unselect the unit
> 
>     * Unit selected + right-click on a unit -> unselect unit and **show context menu** (what this issue is complaining about)
> 
> 
> After [#11368](https://github.com/wesnoth/wesnoth/pull/11368) (and behaviour for 1.12 and 1.10 and likely earlier):
> 
>     * No unit selected + right-click anywhere -> context menu
> 
>     * Unit selected + right-click anywhere except a hex with a unit -> unselect the unit
> 
>     * Unit selected + right-click on a unit -> unselect unit (ie **No more context menu**)

That makes it obvious, thanks.

My issue is not with deselection. The point is some behavior such as custom menu items may need to show in that hex. The empty hex behaviour is also problematic in this case.

My addition:
 * Regardless of the right click location, in addition to the above, context menu must be shown if it has a custom menu item `[set_menu_item]` that triggers at that hex or if right clicking on that hex satisfies it's view condition.

**Comment by soliton-** [comment]
> My issue is not with deselection.

No need to worry about this issue or its solution then.

> Regardless of the right click location, in addition to the above, context menu must be shown if it has a custom menu item `[set_menu_item]` that triggers at that hex or if right clicking on that hex satisfies it's view condition.

Nothing there changes at all if deselection is not involved. Right click works as before. The only change is that if a unit was selected then the first right click deselects and the second works as always.

**Comment by babaissarkar** [comment]
> Nothing there changes at all if deselection is not involved. Right click works as before. The only change is that if a unit was selected then the first right click deselects and the second works as always.

In addition to what @CelticMinstrel already said in #11368, this introduces a new UX issue: user has no way to know they need to click again to show the context menu in certain cases. That's not better than actually checking if a custom menu item exists, and deciding accordingly.

If this is done, there needs to be a way to disable this on the UMC level or it'll be annoying for a new user, one who isn't fully certain of wesnoth selection. (Sidebar and reachmap are the more obvious visual indicators, and they don't really turn off just with deselection.)

**Comment by babaissarkar** [comment]
> What mass UMC rewrite? Are you implying there was a mass UMC rewrite after the behavior change (for the specific case of right clicking on a unit while another is selected as is the only thing this issue is about) of deselecting and opening the context menu? What would there even have been rewritten? D

Not all changes are reversible. Sometimes the forward change requires no rewrite, but the reversion does. Like the UMC disable option I said above. This present dual  behaviour had already existed for multiple releases, creating inertia.

**Comment by soliton-** [comment]
Since there was confusion about some technical issues and breakage I'd like to cite celmin from IRC:
> I don't fully undertand the situation TBH. I haven't done any testing and it's not clear to me what the current situation is. I have a case that _would_ break if a certain understanding of the situation is accurate, but I don't know if that understanding is accurate.

------

The previous behavior was IMO more consistent and clearly worked fine for many years. As new requirements are popping up I'm going to bow out of the conversation since it seems Wedge was spot on with the big debate assessment. Not worth wasting more time on this trivial issue.

**Comment by Pentarctagon** [comment]
To add my own two cents to this debate: I think the newer behavior has been around long enough at this point that changing it would just be another disruption (as small as it is) to what players now consider to be the normal behavior. If there are a bunch of edge cases that need to be sorted out, I think it'd be best to close this issue and open new issues for how it should work in those other situations.

**Comment by gfgtdf** [comment]
from my current understanding of this issue, i'm wondering how needs_select worked before 1.13? Does anyone know?

**Comment by soliton-** [comment]
> from my current understanding of this issue, i'm wondering how needs_select worked before 1.13? Does anyone know?

It worked the same as now and as however many years before. It saves the last select event (while this issue has nothing to do with a select event).

**Comment by babaissarkar** [comment]
Some interesting tidbit I noticed: I was working on #11407 which migrates the event handler in question (still draft, far away), and the default gui2 behavior is hotkey bound action (deselect hex) takes priority over inbuilt/hardcoded action (right click context menu). If the action fails (nothing to deselect) then the inbuilt action (right click menu) is triggered. Which is exactly what this issue was asking, so I won't object as strongly. However, I still think the edge cases should be considered and player/modder opinion taken into account if possible, and I'm willing to implement the alternative (deselect + show context menu) based on consensus.
